.github/workflows/main.yml

```diff
@@ -16,7 +19,8 @@ jobs:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
         python-version: ['3.8', '3.9', '3.10', '3.11']
-      fail-fast: False
+      # fail fast only on pull-requests!
+      fail-fast: ${{ github.event_name == 'pull_request' }}
```

Comment:
Why this change? The CI often fails on Windows but not Linux or on some python versions and not others, so it's very useful to let it all run.

Reply:
Ok I wanted to reduce redundant failures, e.g. because PRs forgot to format or because of some trailing whitespace. But I can remove that.

Reply:
That's a fair point, but probably better handled by just separating the static checks (`flake8`, `mypy`) and only running `pytest` if those succeed.

papis tests are pretty fast, so we never bothered too much :\

Reply:
ok I separated them. Now the lints are run first on `ubuntu-latest` with latest python in a separate step before running all tests/lints with the matrix job.